### PR TITLE
[FLINK-7854] [table] Reject lateral table outer joins with predicates in SQL

### PR DESCRIPTION
**Pull request by xccui**
## What is the purpose of the change

Due to CALCITE-2004, this PR aims to temporarily restrict the predicates for left outer join against the lateral table to be literal `TRUE`.

## Brief change log

  - Override the `validateJoin()` method in `FlinkCalciteSqlValidator` to check the join conditions.
  - Update the tests.
  - Refine the documents.

## Verifying this change

This change is already covered by existing tests `CorrelateTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/FlinkCalciteSqlValidator.scala [hunk 1]
```diff
@@ -48,4 +49,19 @@ class FlinkCalciteSqlValidator(
       insert: SqlInsert): RelDataType = {
     typeFactory.asInstanceOf[JavaTypeFactory].toSql(targetRowType)
   }
+
+  override def validateJoin(join: SqlJoin, scope: SqlValidatorScope): Unit = {
+    // Due to the improperly translation of lateral table left outer join (see CALCITE-2004), we
+    // need to temporarily forbid the common predicates until the problem is fixed.
+    // The check for join with a lateral table is actually quite tricky.
+    if (join.getJoinType == JoinType.LEFT &&
+      join.getRight.toString.startsWith("TABLE(")) { // TABLE (`func`(`foo`)) AS...
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/FlinkCalciteSqlValidator.scala]: I don't free comfortable to check the string representation here.
There should be a better way, maybe recursively checking for `SqlKind.COLLECTION_TABLE`:

```
private def isCollectionTable(node: SqlNode): Boolean = {
    if (node.getKind == SqlKind.COLLECTION_TABLE) {
      true
    } else {
      node match {
        case c: SqlCall => c.getOperandList.asScala.filter(_ != null).exists(o => isCollectionTable(o))
        case l: SqlNodeList => l.getList.asScala.filter(_ != null).exists(o => isCollectionTable(o))
        case _ => false
      }
    }
  }
```

This method passes all tests, but I'm not 100% that it is correct and won't break for more complex queries.
Maybe there's also some methods in `SqlValidatorImpl` that are helpful.

**Comment by fhueske** [comment]
Hi @xccui, are you working on this right now? 
FLINK-7854 is a blocker for the release and we should try to fix it before end of next week.

Thanks, Fabian

**Comment by xccui** [comment]
Hi @fhueske. I am working on it now.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/FlinkCalciteSqlValidator.scala (hunk 1 above)
Comment (xccui) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/FlinkCalciteSqlValidator.scala]: Yes, checking the string here may not be reliable and we need to search for the node whose kind is `SqlKind.COLLECTION_TABLE`.

However, the recursive check could fail with subqueries. For instance, given the following query,
```
 SELECT *
 FROM MyTable2 LEFT OUTER JOIN
  (SELECT c, s
   FROM MyTable LEFT OUTER JOIN LATERAL TABLE(func1(c)) AS T(s) on true)
 ON c2 = s 
```
the lateral table in the subquery will be detected by the out join. Thus I think we should check the collection table, which should always be covered by an `AS` node, directly, i.e.,
```
private def isCollectionTable(node: SqlNode): Boolean = {
  node match {
    case n: SqlCall if n.getKind == SqlKind.AS =>
      n.getOperandList.get(0).getKind == SqlKind.COLLECTION_TABLE
  }
}
```
What do you think?

Comment (fhueske) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/FlinkCalciteSqlValidator.scala]: Yes, you are right.
Let's go for your proposal :-)
Thanks!

**Comment by xccui** [comment]
Hi @fhueske, the PR has been updated.

**Comment by fhueske** [comment]
Great thanks a lot! The PR looks good. Will merge it.

**Pull request closed without merging**